File: 2025/day10/factory.zig
const std = @import("std");

const EPS = 5 * std.math.floatEps(f32);

pub fn parse_diagram(word: []const u8, ndigit: *u5) u32 {
    var res: u32 = 0;
    ndigit.* = 0;
    for (word[0 .. word.len - 1]) |w| {
        ndigit.* += 1;
        res <<= 1;
        if (w == '#') {
            res += 1;
        }
    }
    return res;
}

pub fn parse_button(allocator: std.mem.Allocator, word: []const u8) ![]u8 {
    var array = std.array_list.Managed(u8).init(allocator);
    for (word) |w| {
        if (w == ',' or w == ')') {
            continue;
        }
        try array.append(w - '0');
    }
    return array.toOwnedSlice();
}

pub fn button_to_mask(indices: []const u8, ndigit: u5) u32 {
    var res: u32 = 0;
    for (indices) |i| {
        const n: u5 = @intCast(i);
        res |= @as(u32, 1) << (ndigit - 1 - n);
    }
    return res;
}

pub fn parse_requirements(allocator: std.mem.Allocator, word: []const u8) ![]u16 {
    var jolt = std.array_list.Managed(u16).init(allocator);
    var it = std.mem.tokenizeAny(u8, word[0 .. word.len - 1], ",");
    while (it.next()) |part| {
        const n = try std.fmt.parseInt(u16, part, 10);
        try jolt.append(n);
    }
    return try jolt.toOwnedSlice();
}

const Machine = struct {
    ndigit: u5,
    diagram: u32,
    buttons: [][]u8,
    requirements: []u16,
    allocator: std.mem.Allocator,

    pub fn from_line(allocator: std.mem.Allocator, line: []u8) !Machine {
        var it = std.mem.tokenizeAny(u8, line, " ");
        var diagram: u32 = undefined;
        var jolt: []u16 = undefined;
        var ndigit: u5 = 0;
        var buttons = std.array_list.Managed([]u8).init(allocator);
        while (it.next()) |part| {
            switch (part[0]) {
                '[' => {
                    diagram = parse_diagram(part[1..], &ndigit);
                },
                '(' => {
                    try buttons.append(try parse_button(allocator, part[1..]));
                },
                '{' => {
                    jolt = try parse_requirements(allocator, part[1..]);
                },
                else => unreachable,
            }
        }
        return Machine{ .ndigit = ndigit, .diagram = diagram, .buttons = try buttons.toOwnedSlice(), .requirements = jolt, .allocator = allocator };
    }

    pub fn free(self: *Machine) void {
        for (self.buttons) |v| {
            self.allocator.free(v);
        }
        self.allocator.free(self.buttons);
        self.allocator.free(self.requirements);
    }

    pub fn constraint_matrix(self: *const Machine) ![][]f32 {
        const m = self.requirements.len;
        const n = self.buttons.len;
        var a = try self.allocator.alloc([]f32, m);
        for (0..m) |i| {
            a[i] = try self.allocator.alloc(f32, n);
        }
        for (self.buttons, 0..) |button, j| {
            for (button) |b| {
                const i: usize = @intCast(b);
                a[i][j] = 1.0;
            }
        }
        return a;
    }

    pub fn joltage_vector(self: *const Machine) ![]f32 {
        var b = try self.allocator.alloc(f32, self.requirements.len);
        for (self.requirements, 0..) |v, i| {
            b[i] = @as(f32, @floatFromInt(v));
        }
        return b;
    }

    pub fn check_solution(self: *Machine, x: []const f32, sol: *u32) !bool {
        var jolt = try self.allocator.alloc(u16, self.requirements.len);
        defer self.allocator.free(jolt);
        @memset(jolt, 0);
        for (self.buttons, 0..) |button, i| {
            const ix: i32 = @intFromFloat(@round(x[i]));
            if (ix < 0) return false;
            sol.* += @intCast(ix);
            for (button) |ib| {
                jolt[@intCast(ib)] += @intCast(ix);
            }
        }
        for (self.requirements, 0..) |target, i| {
            if (jolt[i] != target) return false;
        }
        return true;
    }
};

pub fn solve_part1(target: u32, buttons: []const u32, current: u32, cost: u8, best: *u8) void {
    if (current == target) {
        if (cost < best.*) {
            best.* = cost;
        }
        return;
    }
    if (cost > best.* or buttons.len == 0) return;
    solve_part1(target, buttons[1..], current, cost, best);
    solve_part1(target, buttons[1..], current ^ buttons[0], cost + 1, best);
}

pub fn show_matrix(comptime T: type, mat: [][]T) void {
    for (mat) |row| {
        for (row) |v| {
            std.debug.print(" {d:1.3} ", .{v});
        }
        std.debug.print("\n", .{});
    }
}

pub fn free_matrix(allocator: std.mem.Allocator, comptime T: type, mat: [][]T) void {
    for (mat) |row| {
        allocator.free(row);
    }
    allocator.free(mat);
}

pub fn square_matrix(allocator: std.mem.Allocator, comptime T: type, a: [][]T) ![][]T {
    const m = a.len;
    const n = a[0].len;
    var c = try allocator.alloc([]T, n);
    for (0..n) |i| {
        c[i] = try allocator.alloc(T, n);
        @memset(c[i], 0.0);
    }
    for (0..n) |i| {
        for (0..i + 1) |j| {
            for (0..m) |k| {
                const incr = a[k][i] * a[k][j];
                c[i][j] += incr;
                c[j][i] += incr;
            }
            if (i == j) {
                c[i][j] *= 0.5;
            }
        }
    }
    return c;
}

pub fn pre_mult(allocator: std.mem.Allocator, comptime T: type, b: []T, a: [][]T) ![]T {
    std.debug.assert(a.len == b.len);
    var c = try allocator.alloc(T, a[0].len);
    for (0..a[0].len) |i| {
        c[i] = 0;
        for (0..a.len) |j| {
            c[i] += a[j][i] * b[j];
        }
    }
    return c;
}

pub fn dot_product(comptime T: type, vec1: []T, vec2: []T) T {
    std.debug.assert(vec1.len == vec2.len);
    var scalar: T = 0;
    for (0..vec1.len) |i| {
        scalar += vec1[i] * vec2[i];
    }
    return scalar;
}

pub fn least_squares(allocator: std.mem.Allocator, a: [][]f32, b: []f32, atol: f32) ![]f32 {
    const aa = try square_matrix(allocator, f32, a);
    defer free_matrix(allocator, f32, aa);
    const ab = try pre_mult(allocator, f32, b, a);
    defer allocator.free(ab);
    var x = try allocator.alloc(f32, ab.len);

    var r = try allocator.alloc(f32, ab.len);
    defer allocator.free(r);
    var p = try allocator.alloc(f32, ab.len);
    defer allocator.free(p);
    var ap = try allocator.alloc(f32, ab.len);
    defer allocator.free(ap);

    @memcpy(x, ab);
    for (0..ab.len) |i| {
        r[i] = ab[i];
        for (0..ab.len) |j| {
            r[i] -= aa[i][j] * x[j];
        }
    }

    @memcpy(p, r);
    var r2: f32 = dot_product(f32, r, r);
    var r2new: f32 = undefined;
    var delta: f32 = 1000.0;
    while (delta > atol * atol) {
        for (0..ab.len) |i| {
            ap[i] = 0;
            for (0..ab.len) |j| {
                ap[i] += aa[i][j] * p[j];
            }
        }
        const alpha = r2 / dot_product(f32, p, ap);
        for (0..ab.len) |i| {
            x[i] += alpha * p[i];
        }
        // Modify conjugated gradient to make sure that solution is positive
        for (0..ab.len) |i| {
            if (x[i] < 0) x[i] = 0;
            r[i] = ab[i];
            for (0..ab.len) |j| {
                r[i] -= aa[i][j] * x[j];
            }
        }
        r2new = dot_product(f32, r, r);
        for (0..ab.len) |i| {
            p[i] = r[i] + r2new * p[i] / r2;
        }
        delta = (r2 - r2new) * (r2 - r2new);
        r2 = r2new;
    }
    // std.debug.print("residual: {}\n", .{r2});

    return x;
}

pub fn gaussian_elimination(a: [][]f32, b: []f32) void {
    const m = a.len;
    const n = a[0].len;
    var h: usize = 0;
    var k: usize = 0;
    while (h < m and k < n) {
        var imax: usize = h;
        for (h..m) |i| {
            if (@abs(a[i][k]) > @abs(a[imax][k])) {
                imax = i;
            }
        }
        if (std.math.approxEqAbs(f32, a[imax][k], 0, EPS)) {
            k += 1;
            continue;
        }

        std.mem.swap([]f32, &a[imax], &a[h]);
        std.mem.swap(f32, &b[h], &b[imax]);
        for (h + 1..m) |i| {
            const fac = a[i][k] / a[h][k];
            a[i][k] = 0;
            b[i] -= fac * b[h];
            for (k + 1..n) |j| {
                a[i][j] -= fac * a[h][j];
            }
        }
        h += 1;
        k += 1;
    }
}

const System = struct {
    allocator: std.mem.Allocator,
    a: [][]f32,
    b: []f32,
    free_var: []bool,
    upper_bounds: []usize,

    pub fn from_machine(machine: *const Machine) !System {
        const allocator = machine.allocator;
        const a = try machine.constraint_matrix();
        const b = try machine.joltage_vector();
        var upper = try allocator.alloc(usize, a[0].len);
        for (machine.buttons, 0..) |button, ivar| {
            var ub: usize = std.math.maxInt(usize);
            for (button) |ib| {
                const i: usize = @intCast(ib);
                if (machine.requirements[i] < ub) {
                    ub = machine.requirements[i];
                }
            }
            upper[ivar] = ub;
        }
        gaussian_elimination(a, b);
        var i: usize = 0;
        var j: usize = 0;

        var isfree = try allocator.alloc(bool, a[0].len);
        @memset(isfree, true);
        while (i < a.len and j < a[0].len) : (i += 1) {
            while (j < a[0].len and std.math.approxEqAbs(f32, a[i][j], 0, EPS)) {
                j += 1;
            }
            if (j < a[0].len) isfree[j] = false;
        }
        // show_matrix(f32, a);
        // std.debug.print("b={any}\n", .{b});
        // std.debug.print("free_var={any}\n", .{isfree});
        // std.debug.print("upper_bound={any}\n", .{upper});
        return System{ .allocator = allocator, .a = a, .b = b, .free_var = isfree, .upper_bounds = upper };
    }

    pub fn solve(self: *const System, x: []f32) void {
        var i: usize = 0;
        for (self.free_var) |skip| {
            if (!skip) i += 1;
        }
        var j: usize = self.a[0].len;
        while (i > 0) : (i -= 1) {
            j -= 1;
            while (self.free_var[j] and j > 0) {
                j -= 1;
            }
            var k: usize = self.a[0].len - 1;
            x[j] = self.b[i - 1];
            while (k > j) : (k -= 1) {
                if (self.free_var[k]) continue;
                x[j] -= x[k] * self.a[i - 1][k];
            }
            x[j] /= self.a[i - 1][j];
        }
    }

    pub fn free(self: *const System) void {
        free_matrix(self.allocator, f32, self.a);
        self.allocator.free(self.b);
        self.allocator.free(self.free_var);
        self.allocator.free(self.upper_bounds);
    }
};

pub fn solve_part2(machine: *Machine, sys: *System, col: usize, x: []f32, best: *u32) !void {
    var ivar = col;
    const nvar = sys.free_var.len;
    while (ivar < nvar and !sys.free_var[ivar]) {
        ivar += 1;
    }
    if (ivar == nvar) {
        sys.solve(x);
        var sol: u32 = 0;
        const valid_sol = try machine.check_solution(x, &sol);
        // std.debug.print("x={any} , valid={}\n", .{x, valid_sol});
        if (valid_sol and sol < best.*) {
            best.* = sol;
        }
        return;
    }
    const bold = try sys.allocator.alloc(f32, sys.b.len);
    defer sys.allocator.free(bold);
    @memcpy(bold, sys.b);
    for (0..sys.upper_bounds[ivar] + 1) |value| {
        x[ivar] = @floatFromInt(value);
        for (0..sys.a.len) |i| {
            sys.b[i] = bold[i] - sys.a[i][ivar] * x[ivar];
        }
        try solve_part2(machine, sys, ivar + 1, x, best);
    }
}

pub fn main() !void {
    if (std.os.argv.len != 2) {
        return error.InvalidArgument;
    }
    const cwd = std.fs.cwd();
    const file_name: [:0]const u8 = std.mem.span(std.os.argv[1]);
    const file = try cwd.openFile(file_name, .{});
    defer file.close();

    var buffer: [4096]u8 = undefined;
    var reader = file.reader(&buffer);

    const allocator = std.heap.c_allocator;
    var part1: u32 = 0;
    var part2: u32 = 0;

    while (try reader.interface.takeDelimiter('\n')) |line| {
        var machine = try Machine.from_line(allocator, line);
        defer machine.free();
        var buttons: []u32 = try allocator.alloc(u32, machine.buttons.len);
        defer allocator.free(buttons);
        var sol1: u8 = std.math.maxInt(u8);
        for (machine.buttons, 0..) |button, i| {
            buttons[i] = button_to_mask(button, machine.ndigit);
        }
        solve_part1(machine.diagram, buttons, 0, 0, &sol1);
        part1 += sol1;

        var sys = try System.from_machine(&machine);
        defer sys.free();

        var sol2: u32 = std.math.maxInt(u32);
        const x = try allocator.alloc(f32, sys.a[0].len);
        defer allocator.free(x);
        try solve_part2(&machine, &sys, 0, x, &sol2);
        // std.debug.print("-> sol2={}\n", .{sol2});
        part2 += sol2;
    }
    std.debug.print("Part 1: {}\n", .{part1});
    std.debug.print("Part 2: {}\n", .{part2});
}

File: 2025/day10/factory.c
#include <assert.h>
#include <math.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define BUFSIZE 4096
#define ABS(x) (((x) >= 0) ? (x) : -(x))
#define MIN(a, b) (((a) < (b)) ? (a) : (b))

typedef uint8_t u8;
typedef uint16_t u16;
typedef uint32_t u32;
typedef float f32;
typedef int32_t i32;
typedef u32 b32;

typedef struct {
  u32 capacity, length, elem_size;
  void *data;
} Vector;

typedef struct {
  u32 n_digit, diagram;
  Vector *buttons;
  Vector *requirements;
} Machine;

typedef struct {
  u32 shape[2];
  f32 **A;
  f32 *b, *x;
  i32 *xi;
  b32 *isfree;
  u16 *upper;
} System;

Vector *vector_create(u32 capacity, u32 elem_size);
void *vector_get(Vector *self, u32 index);
void *vector_append(Vector *self);
void vector_free(Vector *self);
void vector_empty(Vector *self) { self->length = 0; }

void machine_parse(Machine *machine, char *line);
void machine_display(Machine *self);
b32 machine_button_is_active(Machine *self, u32 req_index, u32 button_index);

void solve_part1(Machine *machine, u32 index, u32 word, u8 cost, u8 *min_cost);

void swapf(f32 *a, f32 *b);
void system_swap_rows(System *sys, u32 i, u32 j);

void system_build(System *sys, Machine *machine);
void system_display(System *sys);
void system_gaussian_elimination(System *sys);
void system_solve(System *sys);
u32 system_integer_solution(System *sys, Machine *machine);
void system_free(System *sys);
void system_print_solution(System *sys);

void solve_part2(Machine *machine, System *sys, u32 col, u32 *min_cost);

int main(int argc, char *argv[]) {
  if (argc < 2) {
    fprintf(stderr, "Missing input file\n");
    return EXIT_FAILURE;
  }

  FILE *fp = fopen(argv[1], "r");
  if (fp == NULL) {
    fprintf(stderr, "Could not open file %s\n", argv[1]);
    return EXIT_FAILURE;
  }

  char line[BUFSIZE];
  Machine machine = {.buttons = vector_create(16, sizeof(u32)),
                     .requirements = vector_create(16, sizeof(u16))};
  u32 part1 = 0, part2 = 0;
  System sys = {};
  while (fgets(line, BUFSIZE, fp)) {
    machine_parse(&machine, line);
    // machine_display(&machine);
    u8 sol1 = UINT8_MAX;
    solve_part1(&machine, 0, 0, 0, &sol1);
    part1 += sol1;

    system_build(&sys, &machine);
    system_gaussian_elimination(&sys);
    u32 sol2 = UINT32_MAX;
    solve_part2(&machine, &sys, 0, &sol2);
    part2 += sol2;

    system_free(&sys);
  }
  fclose(fp);
  printf("Part 1: %u\n", part1);
  printf("Part 2: %u\n", part2);

  vector_free(machine.buttons);
  vector_free(machine.requirements);
  return EXIT_SUCCESS;
}

Vector *vector_create(u32 capacity, u32 elem_size) {
  Vector *vec = malloc(sizeof(Vector));
  vec->capacity = capacity;
  vec->length = 0;
  vec->elem_size = elem_size;
  vec->data = malloc(elem_size * capacity);
  return vec;
}

void *vector_get(Vector *self, u32 index) {
  assert(index < self->length);
  return self->data + self->elem_size * index;
}

void *vector_append(Vector *self) {
  if (self->length == self->capacity) {
    self->capacity *= 2;
    self->data = realloc(self->data, self->capacity * self->elem_size);
  }
  return self->data + (self->length++) * self->elem_size;
}

void vector_free(Vector *self) {
  free(self->data);
  free(self);
}

void machine_parse(Machine *machine, char *line) {
  vector_empty(machine->buttons);
  vector_empty(machine->requirements);
  u32 n = strcspn(line, "]");
  u32 diagram = 0;
  u32 n_digit = 0;
  for (u32 i = 1; i < n; ++i) {
    n_digit++;
    diagram <<= 1;
    if (line[i] == '#')
      diagram += 1;
  }
  machine->n_digit = n_digit;
  machine->diagram = diagram;
  u32 stop = strcspn(line, "{") - 1;
  u32 button = 0;
  for (u32 i = n + 2; i < stop; ++i) {
    if (line[i] == '(' || line[i] == ' ' || line[i] == ',') {
      continue;
    }
    if (line[i] == ')') {
      u32 *appended = (u32 *)vector_append(machine->buttons);
      *appended = button;
      button = 0;
      continue;
    }
    u8 num = line[i] - '0';
    button |= 1 << (n_digit - 1 - num);
  }
  n = strcspn(line, "}");
  u16 num = 0;
  for (u32 i = stop + 2; i <= n; ++i) {
    if (line[i] == ',' || i == n) {
      u16 *ptr = (u16 *)vector_append(machine->requirements);
      *ptr = num;
      num = 0;
      continue;
    }
    num = num * 10 + (line[i] - '0');
  }
}

b32 machine_button_is_active(Machine *self, u32 req_index, u32 button_index) {
  u32 button = *(u32 *)vector_get(self->buttons, button_index);
  u32 test = 1 << (self->n_digit - req_index - 1);
  return button & test;
}

void machine_display(Machine *self) {
  printf("Buttons:");
  for (u32 ibut = 0; ibut < self->buttons->length; ++ibut) {
    printf("(");
    for (u32 ireq = 0; ireq < self->n_digit; ++ireq) {
      if (machine_button_is_active(self, ireq, ibut))
        printf("%u,", ireq);
    }
    printf(") ");
  }
  printf("\nRequirements: {");
  for (u32 i = 0; i < self->requirements->length; ++i) {
    u16 num = *(u16 *)vector_get(self->requirements, i);
    printf("%hu, ", num);
  }
  printf("}\n");
}

void solve_part1(Machine *machine, u32 index, u32 word, u8 cost, u8 *min_cost) {
  if (word == machine->diagram) {
    if (cost < *min_cost)
      *min_cost = cost;
    return;
  }
  if (cost >= *min_cost || index >= machine->buttons->length)
    return;
  u32 button = *(u32 *)vector_get(machine->buttons, index);
  solve_part1(machine, index + 1, word, cost,
              min_cost); // button is not pressed
  solve_part1(machine, index + 1, word ^ button, cost + 1,
              min_cost); // button is pressed;
}

void system_build(System *sys, Machine *machine) {
  sys->shape[1] = machine->buttons->length;
  sys->shape[0] = machine->n_digit;
  sys->b = malloc(sizeof(f32) * sys->shape[0]);
  sys->x = malloc(sizeof(f32) * sys->shape[1]);
  u16 *requirements = (u16 *)machine->requirements->data;
  u32 *buttons = (u32 *)machine->buttons->data;
  sys->upper = malloc(sizeof(u16) * sys->shape[1]);
  for (u32 i = 0; i < sys->shape[1]; ++i) {
    sys->upper[i] = UINT16_MAX;
  }
  f32 *a = malloc(sizeof(f32) * sys->shape[0] * sys->shape[1]);
  for (u32 i = 0; i < sys->shape[0]; ++i) {
    sys->b[i] = (f32)requirements[i];
    for (u32 j = 0; j < sys->shape[1]; ++j) {
      if (machine_button_is_active(machine, i, j)) {
        a[i * sys->shape[1] + j] = 1.0;
        sys->upper[j] = MIN(sys->upper[j], requirements[i]);
      } else {
        a[i * sys->shape[1] + j] = 0.0;
      }
    }
  }
  sys->A = malloc(sizeof(f32 *) * sys->shape[0]);
  sys->A[0] = a;
  for (u32 i = 1; i < sys->shape[0]; ++i) {
    sys->A[i] = sys->A[i - 1] + sys->shape[1];
  }
  sys->isfree = malloc(sizeof(b32) * sys->shape[1]);
  sys->xi = malloc(sizeof(i32) * sys->shape[1]);
}

void system_display(System *sys) {
  printf("A=\n");
  for (u32 i = 0; i < sys->shape[0]; ++i) {
    for (u32 j = 0; j < sys->shape[1]; ++j) {
      printf(" %1.2f ", sys->A[i][j]);
    }
    printf("\n");
  }
  printf("\nb=");
  for (u32 i = 0; i < sys->shape[0]; ++i) {
    printf(" %1.2f ", sys->b[i]);
  }
  printf("\n");
}

void system_print_solution(System *sys) {
  printf("x=");
  for (u32 i = 0; i < sys->shape[1]; ++i) {
    printf(" %1.2f ", sys->x[i]);
  }
  printf("\n");
  printf("xi=");
  for (u32 i = 0; i < sys->shape[1]; ++i) {
    printf(" %d ", sys->xi[i]);
  }
  printf("\n");
}

void system_free(System *sys) {
  free(sys->x);
  free(sys->xi);
  free(sys->b);
  free(sys->A[0]);
  free(sys->A);
  free(sys->isfree);
  free(sys->upper);
}

void swapf(f32 *a, f32 *b) {
  f32 tmp = *b;
  *b = *a;
  *a = tmp;
}

void system_swap_rows(System *sys, u32 i, u32 j) {
  if (i == j)
    return;
  for (u32 k = 0; k < sys->shape[1]; ++k) {
    swapf(&sys->A[i][k], &sys->A[j][k]);
  }
  swapf(&sys->b[i], &sys->b[j]);
}

void system_gaussian_elimination(System *sys) {
  u32 h = 0, k = 0;
  while (h < sys->shape[0] && k < sys->shape[1]) {
    u32 imax = h;
    for (u32 i = h + 1; i < sys->shape[0]; ++i) {
      if (ABS(sys->A[i][k]) > ABS(sys->A[imax][k]))
        imax = i;
    }
    if (ABS(sys->A[imax][k]) < 1e-6) {
      k++;
      continue;
    }
    system_swap_rows(sys, imax, h);
    for (u32 i = h + 1; i < sys->shape[0]; ++i) {
      f32 fac = sys->A[i][k] / sys->A[h][k];
      sys->A[i][k] = 0.0;
      sys->b[i] -= fac * sys->b[h];
      for (u32 j = k + 1; j < sys->shape[1]; ++j) {
        sys->A[i][j] -= fac * sys->A[h][j];
      }
    }
    h++;
    k++;
  }
  for (u32 i = 0; i < sys->shape[1]; ++i) {
    sys->isfree[i] = true;
  }
  u32 j = 0;
  for (u32 i = 0; i < sys->shape[0] && j < sys->shape[1]; ++i) {
    while (j < sys->shape[1] && ABS(sys->A[i][j]) < 1e-6) {
      j++;
    }
    if (j < sys->shape[1])
      sys->isfree[j] = false;
  }
}

void system_solve(System *sys) {
  u32 n = 0;
  for (u32 i = 0; i < sys->shape[1]; ++i) {
    if (!sys->isfree[i]) {
      n++;
    }
  }
  u32 j = sys->shape[1];
  for (u32 i = n; i > 0; --i) {
    j--;
    while (sys->isfree[j] && j > 0) {
      j--;
    }
    sys->x[j] = sys->b[i - 1];
    for (u32 k = sys->shape[1] - 1; k > j; --k) {
      if (sys->isfree[k])
        continue;
      sys->x[j] -= sys->A[i - 1][k] * sys->x[k];
    }
    sys->x[j] /= sys->A[i - 1][j];
  }
  for (u32 i = 0; i < sys->shape[1]; ++i) {
    sys->xi[i] = (i32)roundf(sys->x[i]);
  }
}

u32 system_integer_solution(System *sys, Machine *machine) {
  u16 *target = (u16 *)machine->requirements->data;
  for (u32 i = 0; i < sys->shape[1]; ++i) {
    if (sys->xi[i] < 0)
      return UINT32_MAX;
  }
  for (u32 i = 0; i < sys->shape[0]; ++i) {
    u16 num = 0;
    for (u32 j = 0; j < sys->shape[1]; ++j) {
      if (machine_button_is_active(machine, i, j)) {
        num += (u16)sys->xi[j];
      }
    }
    if (num != target[i])
      return UINT32_MAX;
  }
  u32 isol = 0;
  for (u32 i = 0; i < sys->shape[1]; ++i) {
    isol += (u32)sys->xi[i];
  }
  return isol;
}

void solve_part2(Machine *machine, System *sys, u32 col, u32 *min_cost) {
  u32 ix = col;
  while (ix < sys->shape[1] && !sys->isfree[ix]) {
    ix++;
  }
  if (ix == sys->shape[1]) {
    system_solve(sys);
    // system_print_solution(sys);
    u32 sol = system_integer_solution(sys, machine);
    if (sol < *min_cost) {
      *min_cost = sol;
    }
    return;
  }
  f32 *b = malloc(sizeof(f32) * sys->shape[0]);
  memcpy(b, sys->b, sizeof(f32) * sys->shape[0]);
  for (u16 x_value = 0; x_value <= sys->upper[ix]; ++x_value) {
    sys->x[ix] = (f32)x_value;
    for (u32 i = 0; i < sys->shape[0]; ++i) {
      sys->b[i] = b[i] - sys->A[i][ix] * sys->x[ix];
    }
    solve_part2(machine, sys, ix + 1, min_cost);
  }
  free(b);
}
